--- src/wallpaper/wallpaper.c	2010-03-29 07:09:18.000000000 -0700
+++ src/wallpaper/wallpaper.c	2010-12-28 08:09:46.468000251 -0800
@@ -34,6 +34,8 @@
 
 static int WallpaperDisplayPrivateIndex;
 
+Damage damage;
+
 typedef struct _WallpaperBackground
 {
     char           *image;
@@ -375,6 +377,7 @@
 				     CWBackPixel | CWBorderPixel | CWColormap,
 				     &attr);
 
+	damage = XDamageCreate (dpy, s->root , XDamageReportRawRectangles);
     XSetWMProperties (dpy, ws->fakeDesktop, NULL, NULL,
 		      programArgv, programArgc, &xsh, &xwmh, NULL);
 
@@ -391,9 +394,6 @@
     XFixesSetWindowShapeRegion (dpy, ws->fakeDesktop, ShapeInput, 0, 0, region);
 
     XFixesDestroyRegion (dpy, region);
-
-    XMapWindow (dpy, ws->fakeDesktop);
-    XLowerWindow (dpy, ws->fakeDesktop);
 }
 
 static void
@@ -405,6 +405,9 @@
 	XDestroyWindow (s->display->display, ws->fakeDesktop);
 
     ws->fakeDesktop = None;
+
+	if (damage)
+		XDamageDestroy (s->display->display, damage);
 }
 
 static void
@@ -584,6 +587,7 @@
 {
     CompScreen *s;
     WALLPAPER_DISPLAY (d);
+	int damage_event,damage_error;
 
     UNWRAP (wd, d, handleEvent);
     (*d->handleEvent) (d, event);
@@ -593,6 +597,15 @@
     {
 	WALLPAPER_SCREEN (s);
 
+	Display *dpy=s->display->display;
+	XDamageQueryExtension (dpy, &damage_event, &damage_error);
+	if (event->type == damage_event + XDamageNotify){
+		XDamageNotifyEvent *de = (XDamageNotifyEvent*) event;
+		if (s->root  == de->drawable){
+			XMapWindow ( dpy , ws->fakeDesktop);
+			XLowerWindow ( dpy , ws->fakeDesktop);
+		}
+	}
 	if (!s->desktopWindowCount && ws->fakeDesktop == None
 	    && ws->nBackgrounds)
 	    createFakeDesktopWindow (s);
